### PR TITLE
feat(java_common): rudimentary CompilationUnit FileSystem

### PR DESCRIPTION
**Pull request by shahms**
With the advent of Modules, JavaFileManager gets fairly substantially more complicated dealing with the mapping between module-oriented locations and package-oriented locations.  Rather than re-implementing all of that logic ourselves, using FileSystem-derived Paths backed by a compilation unit potentially enables us to reduce the amount of logic in the indexer's FileManager implementation and instead rely on Path-based location management in the system JavacFileManager for most of that work.

Actually doing this is, unfortunately, dependent on requiring the using of JRE9+, but this lays the groundwork for that.

### CODE REVIEW

File: kythe/java/com/google/devtools/kythe/extractors/shared/ExtractorUtils.java [hunk 1]
```diff
@@ -69,7 +69,7 @@
    * @throws ExtractionException
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/extractors/shared/ExtractorUtils.java]: Fix doc

File: kythe/java/com/google/devtools/kythe/platform/java/filemanager/JavaFileStoreBasedFileManager.java [hunk 2]
```diff
@@ -196,25 +187,20 @@ public FileObject getFileForInput(Location location, String packageName, String
   }
 
   @Override
-  public Iterable<? extends File> getLocation(Location location) {
-    return fileManager.getLocation(location);
+  public boolean contains(Location location, FileObject file) throws IOException {
+    if (file instanceof CustomFileObject) {
+      // We don't currently support modules.
+      return !location.isModuleOrientedLocation();
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/java/filemanager/JavaFileStoreBasedFileManager.java]: `&& super.contains(location, file)` ?

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/CompilationUnitFileAttributes.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.devtools.kythe.platform.shared.filesystem;
+
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.FileTime;
+
+/** {@link BasicFileAttributes} implementat for {@link CompilationUnitFileSystem}. */
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/filesystem/CompilationUnitFileAttributes.java]: sp implementation

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  private long position;
+
+  public ByteBufferByteChannel(byte[] data) {
+    buffer = data;
+    position = 0;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return buffer == null ? false : true;
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java]: `buffer != null`

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  @Override
+  public void close() throws IOException {
+    buffer = null;
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    if (buffer == null) {
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java]: `if (!isOpen()) {`

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.SeekableByteChannel;
+
+/** Read-only {@link SeekableByteChannel} backed by byte[] data. */
+class ByteBufferByteChannel implements SeekableByteChannel {
+  private byte[] buffer;
+  private long position;
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java]: position should be an `int`

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  @Override
+  public long position() throws IOException {
+    return position;
+  }
+
+  @Override
+  public SeekableByteChannel position(long newPosition) throws IOException {
+    checkArgument(newPosition > 0);
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java]: `>=`

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/CompilationUnitPath.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  @Override
+  public boolean startsWith(String other) {
+    return path.startsWith(other);
+  }
+
+  @Override
+  public boolean endsWith(Path other) {
+    return getFileSystem() == other.getFileSystem() ? path.endsWith(unwrap(other)) : false;
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/filesystem/CompilationUnitPath.java]: `getFileSystem() == other.getFileSystem() && path.endsWith(unwrap(other))`

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/CompilationUnitPath.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  @Override
+  public Path subpath(int beginIndex, int endIndex) {
+    return wrap(path.subpath(beginIndex, endIndex));
+  }
+
+  @Override
+  public boolean startsWith(Path other) {
+    return getFileSystem() == other.getFileSystem() ? path.startsWith(unwrap(other)) : false;
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/filesystem/CompilationUnitPath.java]: `getFileSystem() == other.getFileSystem() && path.startsWith(unwrap(other))`

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java (hunk 4 above)
Comment (shahms) [reply kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java]: lol, thanks.

File: kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java (hunk 6 above)
Comment (shahms) [reply kythe/java/com/google/devtools/kythe/platform/shared/filesystem/ByteBufferByteChannel.java]: It shouldn't. The API is specified in terms of long and the mixed math between int and long is isolated to the `read` function.

**Comment by shahms** [comment]
PTAL

**Pull request merged**